File: sdk-swift/Sources/DecibelSDK/Models/Market.swift
import Foundation

public struct PerpMarketConfig: Codable {
    public let marketAddr: String
    public let marketName: String
    public let szDecimals: Int32
    public let pxDecimals: Int32
    public let maxLeverage: Double
    public let minSize: Double
    public let lotSize: Double
    public let tickSize: Double
    public let maxOpenInterest: Double
    public let marginCallFeePct: Double
    public let takerInNextBlock: Bool

    enum CodingKeys: String, CodingKey {
        case marketAddr = "market_addr"
        case marketName = "market_name"
        case szDecimals = "sz_decimals"
        case pxDecimals = "px_decimals"
        case maxLeverage = "max_leverage"
        case minSize = "min_size"
        case lotSize = "lot_size"
        case tickSize = "tick_size"
        case maxOpenInterest = "max_open_interest"
        case marginCallFeePct = "margin_call_fee_pct"
        case takerInNextBlock = "taker_in_next_block"
    }
}

public struct MarketDepth: Codable {
    public let market: String
    public let bids: [MarketOrder]
    public let asks: [MarketOrder]
    public let unixMs: Int64

    enum CodingKeys: String, CodingKey {
        case market, bids, asks
        case unixMs = "unix_ms"
    }
}

public struct MarketOrder: Codable {
    public let price: Double
    public let size: Double
}

public struct MarketPrice: Codable {
    public let market: String
    public let markPx: Double
    public let midPx: Double
    public let oraclePx: Double
    public let fundingRateBps: Double
    public let isFundingPositive: Bool
    public let openInterest: Double
    public let transactionUnixMs: Int64

    enum CodingKeys: String, CodingKey {
        case market
        case markPx = "mark_px"
        case midPx = "mid_px"
        case oraclePx = "oracle_px"
        case fundingRateBps = "funding_rate_bps"
        case isFundingPositive = "is_funding_positive"
        case openInterest = "open_interest"
        case transactionUnixMs = "transaction_unix_ms"
    }
}

public struct MarketContext: Codable {
    public let market: String
    public let volume24h: Double
    public let openInterest: Double
    public let previousDayPrice: Double
    public let priceChangePct24h: Double

    enum CodingKeys: String, CodingKey {
        case market
        case volume24h = "volume_24h"
        case openInterest = "open_interest"
        case previousDayPrice = "previous_day_price"
        case priceChangePct24h = "price_change_pct_24h"
    }
}

public struct Candlestick: Codable {
    public let closeTimestamp: Int64
    public let close: Double
    public let high: Double
    public let interval: String
    public let low: Double
    public let open: Double
    public let openTimestamp: Int64
    public let volume: Double

    enum CodingKeys: String, CodingKey {
        case closeTimestamp = "T"
        case close = "c"
        case high = "h"
        case interval = "i"
        case low = "l"
        case open = "o"
        case openTimestamp = "t"
        case volume = "v"
    }
}

public struct MarketTrade: Codable {
    public let market: String
    public let price: Double
    public let size: Double
    public let isBuy: Bool
    public let unixMs: Int64

    enum CodingKeys: String, CodingKey {
        case market, price, size
        case isBuy = "is_buy"
        case unixMs = "unix_ms"
    }
}

File: sdk-go/models/market.go
package models

// PerpMarketConfig holds the configuration of a perpetual market.
type PerpMarketConfig struct {
	MarketAddr       string  `json:"market_addr"`
	MarketName       string  `json:"market_name"`
	SzDecimals       int32   `json:"sz_decimals"`
	PxDecimals       int32   `json:"px_decimals"`
	MaxLeverage      float64 `json:"max_leverage"`
	MinSize          float64 `json:"min_size"`
	LotSize          float64 `json:"lot_size"`
	TickSize         float64 `json:"tick_size"`
	MaxOpenInterest  float64 `json:"max_open_interest"`
	MarginCallFeePct float64 `json:"margin_call_fee_pct"`
	TakerInNextBlock bool    `json:"taker_in_next_block"`
}

// MarketDepth represents the order book depth.
type MarketDepth struct {
	Market string        `json:"market"`
	Bids   []MarketOrder `json:"bids"`
	Asks   []MarketOrder `json:"asks"`
	UnixMs int64         `json:"unix_ms"`
}

// MarketOrder represents a single order in the book.
type MarketOrder struct {
	Price float64 `json:"price"`
	Size  float64 `json:"size"`
}

// MarketPrice represents price data for a market.
type MarketPrice struct {
	Market            string  `json:"market"`
	MarkPx            float64 `json:"mark_px"`
	MidPx             float64 `json:"mid_px"`
	OraclePx          float64 `json:"oracle_px"`
	FundingRateBps    float64 `json:"funding_rate_bps"`
	IsFundingPositive bool    `json:"is_funding_positive"`
	OpenInterest      float64 `json:"open_interest"`
	TransactionUnixMs int64   `json:"transaction_unix_ms"`
}

// MarketContext provides additional market metadata.
type MarketContext struct {
	Market           string  `json:"market"`
	Volume24h        float64 `json:"volume_24h"`
	OpenInterest     float64 `json:"open_interest"`
	PreviousDayPrice float64 `json:"previous_day_price"`
	PriceChangePct   float64 `json:"price_change_pct_24h"`
}

// Candlestick represents OHLCV data.
type Candlestick struct {
	CloseTimestamp int64   `json:"T"`
	Close          float64 `json:"c"`
	High           float64 `json:"h"`
	Interval       string  `json:"i"`
	Low            float64 `json:"l"`
	Open           float64 `json:"o"`
	OpenTimestamp   int64   `json:"t"`
	Volume         float64 `json:"v"`
}

// MarketTrade represents a trade on a market.
type MarketTrade struct {
	Market string  `json:"market"`
	Price  float64 `json:"price"`
	Size   float64 `json:"size"`
	IsBuy  bool    `json:"is_buy"`
	UnixMs int64   `json:"unix_ms"`
}
